presto-tests/src/test/java/com/facebook/presto/tests/TestMetadataManager.java

```diff
@@ -71,4 +79,30 @@ public void testMetadataIsClearedAfterQueryFailed()
 
         assertEquals(metadataManager.getCatalogsByQueryId().size(), 0);
     }
+
+    @Test(timeOut = 60_000)
+    public void testMetadataIsClearedAfterQueryCanceled()
+            throws Exception
+    {
+        ExecutorService queryExecutor = Executors.newSingleThreadExecutor();
+        DistributedQueryRunner distributedQueryRunner = createQueryRunner();
+        MetadataManager metadata = (MetadataManager) distributedQueryRunner.getMetadata();
+        TestingPrestoServer coordinator = distributedQueryRunner.getCoordinator();
+
+        // submit a relatively long running query
+        queryExecutor.submit(() -> distributedQueryRunner.execute("SELECT COUNT(*) FROM lineitem"));
```

Comment:
I am concerned that this query could finish before the query can be canceled.  Instead I would do something like
```java
        DistributedQueryRunner distributedQueryRunner = createQueryRunner();
        QueryManager queryManager = distributedQueryRunner.getCoordinator().getQueryManager();
        MetadataManager metadata = (MetadataManager) distributedQueryRunner.getMetadata();

        QueryId queryId = queryManager.createQuery(new TestingSessionFactory(TEST_SESSION), "SELECT * FROM lineitem").getQueryId();

        // wait until query starts running                                        
        while (true) {
            if (queryManager.getQueryInfo(queryId).getState() != RUNNING) {
                break;
            }
            Thread.sleep(100);
        }

        // cancel query
        queryManager.cancelQuery(queryId);
        assertEquals(metadata.getCatalogsByQueryId().size(), 0);
```